measure dispatch overhead

zpcore opened the pull request:
Follow up on discussion https://github.com/pytorch/pytorch/pull/159289#discussion_r2245703723. Add the test to detect dtensor dispatch overhead.

The performance number should depend on the specific hardware. Need to figure out a way to enforce the constraint. Worst case is that we can disable the test and let user run locally for manual comparison.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159699



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

zpcore committed:
measure dispatch overhead

[ghstack-poisoned]

wconstab reviewed (commented):
Lgtm! I would like to understand the performance on CI a bit and how you tuned the threshold to make sure it's not going to fail due to noise

wconstab commented on test/distributed/tensor/test_dtensor_dispatch_overhead.py:
```diff
@@ -0,0 +1,133 @@
+            for _ in range(100):
+                # barrier.default() is also an op
+                torch.distributed.barrier()
+                propagator.propagate_op_sharding.cache.cache_clear()
```
I don't see why you use a barrier. There shouldn't be any comms happening in the add anyway so each rank can do the timing locally. Adding the barrier just adds noise right?

wconstab replied:
hmm.
to be clear, we expect the benchmark operation (add) to NOT do any communication inside, so each rank is doing their own timings locally without any interaction, right?

If so, then i think the effects the barrier have are coincidental.

zpcore replied:
I see, it just do the pointwise `add` on local shard repeatly, so there shouldn't be any comm ops.

ezyang commented:
For overhead stuff like this instruction counts are a proxy that is a lot more stable!

zpcore commented on test/distributed/tensor/test_dtensor_dispatch_overhead.py:
```diff
@@ -0,0 +1,133 @@
+        a = distribute_tensor(input_data, device_mesh, [Shard(0)])
+        # warm up
+        with TimeCaptureMode() as tcm:
+            for _ in range(100):
```
Yes!

zpcore commented:
> Lgtm! I would like to understand the performance on CI a bit and how you tuned the threshold to make sure it's not going to fail due to noise

Shall we enable the log message but don't enable the assertTrue check first? Just in case it keeps failing and block others PR. I feel that the number in FB source may also be different depends on the GPU. Once we know more about the performance number, we can enable the assert check.

zpcore committed:
Update on "measure dispatch overhead"

Follow up on discussion https://github.com/pytorch/pytorch/pull/159289#discussion_r2245703723. Add the test to detect dtensor dispatch overhead.

The performance number should depend on the specific hardware. Need to figure out a way to enforce the constraint. Worst case is that we can disable the test and let user run locally for manual comparison.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

wconstab reviewed (approved):
Thanks for the cleanups. I think we can explore instruction counting as a follow up but it would be useful to land this first (with assert disabled if not confident about flakiness)

zpcore committed:
Update on "measure dispatch overhead"

Follow up on discussion https://github.com/pytorch/pytorch/pull/159289#discussion_r2245703723. Add the test to detect dtensor dispatch overhead.

The performance number should depend on the specific hardware. Need to figure out a way to enforce the constraint. Worst case is that we can disable the test and let user run locally for manual comparison.





cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]
